### PR TITLE
[DS-3940] SHERPA API v2 refactor

### PR DESCRIPTION
**Pull request by kshepherd**
Developed by The Library Code with the support of Technische Universität Hamburg

## References
* Fixes #7287
* [JIRA issue link](https://jira.lyrasis.org/browse/DS-3940) 
* [Register for SHERPA v2 API key here](https://v2.sherpa.ac.uk/romeo/api.html)

## Description
The legacy SHERPA ROMeO API that DSpace 6.x (and earlier) is using in both XMLUI and JSPUI for journal deposit policy lookup will be retired in April. 
This refactor uses the new v2 API.
Some changes to display are necessary because of the way policy, journal and publisher data is structured in the new API.

## Instructions for Reviewers

To test this PR:
* Register for a SHERPA API v2 key and configure it in local.cfg
* Begin a submission with a valid ISSN entered into metadata
* At Upload Step, inspect the SHERPA ROMeO policy information and compare it to the source policy page, existing DSpace legacy API response, common sense, etc.
* Try with variant options -- no ISSN, multiple ISSNs, invalid ISSN, mix of valid and invalid ISSNs

List of changes in this PR:
* Deprecated old SHERPA* beans and moved them to org.dspace.app.sherpa.v1
* Wrote new SHERPAResponse class and the beans to represent new data classes in org.dspace.app.sherpa.v2
* Made changes to the servlet and JSP display in JSPUI where appropriate - notably, no more colours, and conditions are grouped by version and option rather than lumped together in "general conditions"
* Made small changes to XMLUI - no visible difference because it just provides an indicator and a link to the ROMeO policy page
* As the v2 API filters don't allow searching for multiple ISSNs, made a change so that just the first valid API response for a set of ISSNs would be used

### CODE REVIEW

**Comment by nwoodward** [comment]
This PR looks really good. I have a few observations and a question. On both the XMLUI and JSPUI submission forms I can enter in the same ISSN multiple times in the Describe Step. If I'm in XMLUI then on the Upload Step I see multiple instances of the same journal in Publisher Information. If I'm in JSPUI, in the Upload Step I only see the journal information for the first ISSN in the list, even if there are multiple valid ISSNs in the Describe Step. 

Is that how things are supposed to work? I checked on a DSpace with the old SHERPA API enabled, and I was able to add the same ISSN multiple times in the Describe Step, so I'm thinking this is probably intentional.

**Comment by kshepherd** [comment]
Thanks @nwoodward 
I haven't changed the way XMLUI works for multiple ISSNs - though I did for JSPUI to just take the first, as you note (the UI approach isn't really set up to display multiple results).

The code comments for XMLUI indicates that displaying indicators/links for *each* journal/ISSN was intentional, though probably not for duplicate ISSNs! That should be a small easy fix.

**Comment by kshepherd** [comment]
> This PR looks really good. I have a few observations and a question. On both the XMLUI and JSPUI submission forms I can enter in the same ISSN multiple times in the Describe Step. If I'm in XMLUI then on the Upload Step I see multiple instances of the same journal in Publisher Information. If I'm in JSPUI, in the Upload Step I only see the journal information for the first ISSN in the list, even if there are multiple valid ISSNs in the Describe Step.
> 
> Is that how things are supposed to work? I checked on a DSpace with the old SHERPA API enabled, and I was able to add the same ISSN multiple times in the Describe Step, so I'm thinking this is probably intentional.

This is now fixed - in my changes, I thought I was simplifying some code by returning a List instead of a Set for the getISSNs method, but it turns out LinkedHashSet was being used specifically to avoid duplicates.

**Review by nwoodward — APPROVED** [review]
:+1: Everything looks good!

**Review by J4bbi — COMMENTED** [review]
I think it would be good if a summary could be provided on the upload step page.
Something like "SHERPA gave 4 results out of 5 ISSNs entered".

Especially on the JSPUI there should be an explanation that the result is only for the first ISSN.

It is not intuitive that if two ISSNs are provided, the first one bogus and the second okay, that a "No results" message is given without any caveat.

File: dspace-jspui/src/main/webapp/sherpa/css/sherpa.css
Comment (kshepherd) [reply dspace-jspui/src/main/webapp/sherpa/css/sherpa.css]: Thanks, I've changed the overflow to auto

File: dspace-api/src/main/resources/Messages.properties
Comment (kshepherd) [reply dspace-api/src/main/resources/Messages.properties:908]: I have changed the JSPUI handling so that each ISSN is now search separately, as with XMLUI. Additional results will be displayed below (since we are scrolling overflow, this is generally OK), as well as adding messages to indicate when there are multiple results, and/or when at least one ISSN returned an error or invalid response

**Comment by nwoodward** [comment]
The additional results display in JSPUI looks good, and I saw the some-errors message as well.

**Review by J4bbi — APPROVED** [review]

**Comment by J4bbi** [comment]
The multiple issn handling in JSPUI looks really good. Nice work.

**Review by KevinVdV — CHANGES_REQUESTED** [review]
@kshepherd I made a review of the code, there are a few things below that I would change.

I also noticed that not everything is parsed yet (which is fine), but it would be best to store the publisher url as it is used in the xmlui, currently the publisher url is linking to the DSpace url on that page.

File: dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java]: Thanks, I have done this in a finally block (and the main methods now throw IOException)

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:107]: Thanks, I have done this in a finally block (and the main methods now throw IOException)

File: dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java]: Removed

File: dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPublisher.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPublisher.java]: Removed

File: dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java]: I've split this method into parts, so there are now separate methods for parsing PublisherPolicy, PermittedVersion, Publisher, Journal, and SystemMetadata. I have had to slightly rework some other logic and models to handle the change in scope, but it is working the same and hopefully is easier to follow. I also added javadoc annotation for each method.

File: dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java]: Removed

**Comment by kshepherd** [comment]
> @kshepherd I made a review of the code, there are a few things below that I would change.
> 
> I also noticed that not everything is parsed yet (which is fine), but it would be best to store the publisher url as it is used in the xmlui, currently the publisher url is linking to the DSpace url on that page.

Hi @KevinVdV , thanks for the review and change requests - I have made all the changes requested

- Some unused and obsolete attributes were removed
- The formatted text (not part of actual response object) was removed
- the parseJSON method was split into several methods to make everything a bit easier to follow and maintain
- javadocs were added
- publisher URL is fixed
- inputstreamreader is always closed

Could you please take another look and update your review? Thanks!

**Comment by kshepherd** [comment]
(note, the Travis CI build appears to be failing because of a Forbidden response from a mvn repo - I'll restart it periodically)

**Comment by kshepherd** [comment]
closing and reopening to force rebuild

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java
Comment (KevinVdV) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:107]: @kshepherd Did you forget to commit the fix here ? Because this class wasn't altered recently.

**Review by KevinVdV — CHANGES_REQUESTED** [review]
@kshepherd Just one more thing that needs to be fixed, you mentioned that you did fix it, so perhaps you forgot to commit it ?

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java]: Thanks, I've pushed a change now

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java
Comment (KevinVdV) [inline dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:112]: Can we place this content.close() in a finally ? If for whatever reason the parsing fails the inputStream won't be closed. After this change I approve of this PR !

Ps: We should also remember to port this PR to the latest DSpace 7 master.

**Review by KevinVdV — APPROVED** [review]

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java]: Good point, I did that in my other implementation, I'm not sure why this one was different.
I have rewritten this small section for a try/catch/finally block, and in the same commit I updated the class to conform to DSpace 7 code style for easier porting.

**Comment by kshepherd** [comment]
Merging as I have review and +1 from a committer, +1 from another committer (though not registered here as we work together) and 2 +1s from trusted devs who are members of release team.
Thanks all for the reviews and improvement suggestions.

**Comment by kshepherd** [comment]
@benpal  yes, this should merge / cherry-pick cleanly into DSpace 6.3, and there is a DSpace CRIS 5.8 branch which I think should be fine to cherry-pick (see the last two commits) into standard DSpace 5.8, at https://github.com/the-library-code/DSpace/commits/SHERPA_API_v2_dspace-cris_5x

**Comment by benpal** [comment]
Thank you so much sir @kshepherd.

**Comment by benpal** [comment]
Hello everyone,

I'm having this error:
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
[ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
[ERROR]   symbol:   method getSizeBytes()
[ERROR]   location: variable bitstream of type org.dspace.content.Bitstream

**Comment by kshepherd** [comment]
> Hello everyone,
> 
> I'm having this error:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
> [ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
> [ERROR] symbol: method getSizeBytes()
> [ERROR] location: variable bitstream of type org.dspace.content.Bitstream

Hi @benpal -  this is an example of the kind of difference between DSpace 5.x and 6.x -- getSize() is the 5.x method and getSizeBytes() is for 6.x

It looks like you're building 6.x here so I'm not sure why it's complaining about this method... is it possible there is a mixup with some source? Pretty sure this method isn't called by any of the commits in this PR.

**Comment by benpal** [comment]
> > Hello everyone,
> > I'm having this error:
> > [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
> > [ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
> > [ERROR] symbol: method getSizeBytes()
> > [ERROR] location: variable bitstream of type org.dspace.content.Bitstream
> 
> Hi @benpal - this is an example of the kind of difference between DSpace 5.x and 6.x -- getSize() is the 5.x method and getSizeBytes() is for 6.x
> 
> It looks like you're building 6.x here so I'm not sure why it's complaining about this method... is it possible there is a mixup with some source? Pretty sure this method isn't called by any of the commits in this PR.

Thank you @kshepherd, Yes I am testing both in 5.8 and 6.3, but for this case, I am using 6.3 as my test case. It seems that the method 'getSizeBytes()' symbol is not recognized by maven compiler.

**Comment by eulereadgbe** [comment]
Hi @kshepherd. It seems the authority control for SHERPARoMEOPublisher and SHERPARoMEOJournalTitle are not working. I tested it both in JSPUI and XMLUI and using suggest and lookup in the choices.presentation. When using lookup, the error that pops up is "Failed to load choice data:  Server indicates error in response." The SHERPA/RoMEO Publisher Policy Database for JSPUI and the Publisher information for XMLUI during the Upload stage however are working correctly. Please advise on how to make the authority control for publisher and journal-title work. Thanks in advance!

**Comment by kshepherd** [comment]
Hi @eulereadgbe - this issue was specifically about fixing the submission-time SHERPA lookups, not the authority plugins, so they'll need updating separately unfortunately. It should be not too difficult to refactor the plugins to use the new v2 classes for Publisher and Journal, but it's not something I've looked at yet.

**Pull request merged**
